NKL/Encode.lean

```diff
@@ -0,0 +1,331 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Govereau
+-/
+import NKL.NKI
+
+/-!
+# Serialization and Deserialization
+
+-/
+
+namespace NKL
+
+-- All of the encode function are pure; decoding uses an instance of EStateM.
+
+abbrev DecodeM := EStateM String ByteArray.Iterator
+
+def decode' (f : DecodeM a) (ba : ByteArray) : Option a :=
+  EStateM.run' f ba.iter
+
+def decode (f : DecodeM a) (ba : ByteArray) : Except String a :=
+  match EStateM.run f ba.iter with
+  | .ok x _ => .ok x
+  | .error s _ => .error s
+
+def decodeFile (f : DecodeM a) (path : System.FilePath) : IO a := do
+ let buf <- IO.FS.readBinFile path
+ match decode f buf with
+ | .ok x => return x
+ | .error s => throw $ IO.userError s
+
+private def next : DecodeM UInt8 := do
+  let it <- get
+  if it.atEnd then throw "unexpected EOF"
+  let b := it.curr
+  set it.next
+  return b
+
+-- TODO: this could be more efficient if we complicate the monad state
+private def take (n : Nat) : DecodeM ByteArray :=
+  let rec f (ba : ByteArray) (n : Nat) : DecodeM ByteArray :=
+    match n with
+    | .zero => return ba
+    | .succ n => next >>= fun b => f (ba.push b) n
+  f (.mk #[]) n
+
+------------------------------------------------------------------------------
+-- Nats are encoded as u64, little-endian
+-- TODO: if the nat is too large, it will be silently truncated.
+
+private def encNat (n : Nat) : ByteArray := .mk
+  #[ UInt8.ofNat $ n >>> 0
+   , UInt8.ofNat $ n >>> 8
+   , UInt8.ofNat $ n >>> 16
+   , UInt8.ofNat $ n >>> 24
+   , UInt8.ofNat $ n >>> 32
+   , UInt8.ofNat $ n >>> 40
+   , UInt8.ofNat $ n >>> 48
+   , UInt8.ofNat $ n >>> 56
+   ]
+
+private def decNat : DecodeM Nat := do
```

Comment:
Could we do `UInt64` here so there is no question about the number of bytes?

Reply:
Yeah, I think I will switch the AST to use Int64 (upgrading to lean 4.14.0)